crates/wasmtime/src/func.rs

```diff
@@ -677,7 +682,33 @@ impl Func {
     ///
     /// Panics if `store` does not own this function.
     pub fn ty(&self, store: impl AsContext) -> FuncType {
-        store.as_context()[self.0].ty.clone()
+        self.load_ty(&store.as_context().0)
+    }
+
+    /// Forcibly loads the type of this function from the `Engine`.
+    ///
+    /// Note that this is a somewhat expensive method since it requires taking a
+    /// lock as well as cloning a type.
+    fn load_ty(&self, store: &StoreOpaque) -> FuncType {
+        FuncType::from_wasm_func_type(
+            store
+                .engine()
+                .signatures()
+                .lookup_type(self.sig_index(store.store_data()))
+                .expect("signature should be registered"),
+        )
+    }
+
+    /// Gets a reference to the `FuncType` for this function.
+    fn ty_ref<'a>(&self, store: &'a mut StoreOpaque) -> (&'a FuncType, &'a StoreOpaque) {
```

Comment:
There's probably a mundane borrow-checker-subtlety reason for returning the downgraded immutable-borrow of `StoreOpaque` here, but could we add a comment describing why?